File: Python3/프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import defaultdict, deque

# 넓이 우선 탐색으로 순회한 노드 수 반환
def bfs(graph, start):
    visited = [start]  
    q = deque([start])  # 시작 노드를 큐에 추가
    n = 1  # 방문한 노드의 수를 초기화

    while q:
        node = q.popleft()  # 큐의 앞에서 노드를 꺼냄
        for adjacent in graph[node]:  # 인접한 모든 노드를 순회
            if adjacent not in visited:  # 인접 노드가 방문되지 않았다면
                visited.append(adjacent)  # 방문한 노드로 표시
                q.append(adjacent)  # 인접 노드를 큐에 추가
                n += 1  # 방문한 노드의 수 증가

    return n  # 방문한 총 노드의 수 반환


def solution(n, wires):
    arr = []  

    for i in wires:  # 자를 각 전선을 순회
        graph = defaultdict(list)  # 그래프를 리스트의 딕셔너리로 초기화
        x, y = i  # 자를 전선의 양 끝 노드를 가져옴
        for j in wires:
            if i == j:  # 현재 자를 전선은 건너뜀
                continue
            a, b = j  # 다른 전선의 양 끝 노드를 가져옴
            # 그래프에 간선을 추가 (무방향)
            graph[a].append(b)  
            graph[b].append(a)  
        print(graph)
        n1 = bfs(graph, x)  
        n2 = bfs(graph, y)  
        arr.append(abs(n1 - n2))  # 두 서브트리 크기의 차이를 리스트에 추가

    answer = min(arr) 

    return answer  

File: Java/프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    
    // BFS로 그래프 순회한 노드 수 반환
    public int bfs(Map<Integer, List<Integer>> graph, int start, boolean[] visited) {
        Queue<Integer> q = new LinkedList<>();
        q.offer(start);  // 시작 노드를 큐에 추가
        visited[start] = true;  // 방문 처리
        int count = 1;  // 방문한 노드 수 초기화

        while (!q.isEmpty()) {
            int node = q.poll();  // 큐에서 노드를 꺼냄
            for (int adjacent : graph.get(node)) {  // 인접한 노드들을 순회
                if (!visited[adjacent]) {  // 방문하지 않은 노드만
                    visited[adjacent] = true;  // 방문 처리
                    q.offer(adjacent);  // 큐에 인접 노드를 추가
                    count++;  // 방문한 노드 수 증가
                }
            }
        }

        return count;  // 방문한 총 노드 수 반환
    }

    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;  // 결과값을 큰 값으로 초기화

        // 각 전선을 하나씩 제거하여 시뮬레이션
        for (int[] wireToCut : wires) {
            Map<Integer, List<Integer>> graph = new HashMap<>();  // 그래프 초기화

            // 그래프 초기화
            for (int i = 1; i <= n; i++) {
                graph.put(i, new ArrayList<>());  // 각 노드의 인접 리스트 생성
            }

            // 전선 그래프 생성
            for (int[] wire : wires) {
                if (Arrays.equals(wire, wireToCut)) {
                    continue;  // 자를 전선은 건너뜀
                }
                int a = wire[0];
                int b = wire[1];
                graph.get(a).add(b);  // 무방향 그래프이므로 양방향 연결
                graph.get(b).add(a);
            }

            // 한 쪽 서브트리 크기 계산 (시작 노드는 자른 전선의 한 쪽 노드)
            boolean[] visited = new boolean[n + 1];  // 노드 방문 여부 체크
            int n1 = bfs(graph, wireToCut[0], visited);  // 첫 번째 노드로부터 BFS 실행
            int n2 = n - n1;  // 전체에서 n1을 제외한 나머지 노드 수

            // 두 서브트리의 크기 차이를 계산
            answer = Math.min(answer, Math.abs(n1 - n2));
        }

        return answer;
    }
}
